# Ejercicio RSA

> Por Mapachana

COnsideramos los bloques $7777$ y $0080$. COmo $0080$ es menor de 1000, rotamos hasta tener $8000$.
Sean $p$ y $q$ los siguientes primos a estos números:

In [1]:
bloq1 = 7777
bloq2 = 8000

p = next_prime(bloq1)
q = next_prime(bloq2)
print(p)
print(q)

7789
8009


$p=7789$ y $q=8009$

Definimos $n=pq$ y $e$ el menor primo mayor o igual que $11$ que es primo relativo con $\varphi(n)$, y $d$ su inverso módulo $\varphi$

In [12]:
n = p*q
print(n)

e = 11
phi=n+1-(p+q)
mcd = e
print(phi)

while(mcd != 1):
    print(e)
    mcd = e.gcd(phi)
    print(mcd)
    e = next_prime(e+1)
    


62382101
62366304
11
11
13
13
17
1


In [14]:
phi%11
phi%13
phi%17

2

In [15]:
e = 17
d = inverse_mod(e,phi)
print(d)
print((e*d)%phi)

29348849
1


Luego tenemos que $e=17$ y $d=29348849$

## Apartado 1

**Cifra el mensaje m = 0xCAFE**

In [17]:
m = 0xcafe
print(m)

51966


In [23]:
c = power_mod(m,e,n)
print(c)
print(hex(c))

55134016
0x3494740


Calculamos $c=m^e \mod n$ obteniendo en hexadecimal 0x3494740

## Apartado 2

**Descifra el criptograma anterior.**

In [24]:
m2 = power_mod(c,d,n)
print(m2)
print(hex(m2))

51966
0xcafe


Calculamos $m= c^d \mod n$ obteniendo así 0xcafe, que era el mensaje original.

## Apartado 3

**Intenta factorizar n mediante el método P − 1 de Pollard. Para
ello llega, como máximo a b = 8**

In [28]:
m = 2
for b in range(1,9):
    print(b)
    pot = power_mod(m,factorial(b),n)
    print(n.gcd(pot-1))
    print("")

1
1

2
1

3
1

4
1

5
1

6
1

7
1

8
1



In [33]:
for m in range(1,1000):
    for b in range(1,9):
        #print(b)
        #print(m)
        pot = power_mod(m,factorial(b),n)
        mcd = n.gcd(pot-1)
        #print(mcd)
        #print("")
        if(mcd != 1):
            print("AAAAAAAAAAAAAAAA")
            print(m)
            print(b)
            print(mcd)
            break

AAAAAAAAAAAAAAAA
1
1
62382101
AAAAAAAAAAAAAAAA
233
3
7789
AAAAAAAAAAAAAAAA
234
3
7789
AAAAAAAAAAAAAAAA
279
7
8009
AAAAAAAAAAAAAAAA
283
4
8009
AAAAAAAAAAAAAAAA
385
4
7789
AAAAAAAAAAAAAAAA
488
7
8009
AAAAAAAAAAAAAAAA
491
7
8009
AAAAAAAAAAAAAAAA
662
7
8009
AAAAAAAAAAAAAAAA
811
7
8009
AAAAAAAAAAAAAAAA
836
7
8009
AAAAAAAAAAAAAAAA
983
7
8009


In [34]:
# Comprobacion algoritmo de Juanmi
parada=false
for i in IntegerRange(5,300):
    if i.gcd(n)==1:
        m=i
        b=1
        a=m.powermod(b,n)
        if((a-1).gcd(n) !=1):
            print (i,a,b)
            print ((a-1).gcd(n))
            parada=true
        if parada==true:
            break
        for b in [2,3,4,5,6,7,8]:
            a=a.powermod(b,n)
            if (a-1).gcd(n)!=1:
                print(i,a,b)
                print('Uno de los factores de n es:') 
                print((a-1).gcd(n))
                parada=true
                break
        if parada==true:
            break

233 4221639 3
Uno de los factores de n es:
7789


In [35]:
factor1=7789
factor2=n//factor1
print(factor1, factor2)
print('Hemos encotrado la descomposición de n en factores primos.')
n==factor1*factor2

7789 8009
Hemos encotrado la descomposición de n en factores primos.


True

Luego obtenemos que con $m=233$ y $b=3$ tenemos que su mcd es $7789$, que es por tanto uno de los factores de $n$, y el otro es por tanto $8009$.

## Apartado 4

**Intenta factorizar $n$ a partir de $\varphi(n)$.**

In [37]:
print(n+1-(p+q))
phi

62366304


62366304

Sabemos que $\varphi(n)=62366304$, luego consideramos $(x-p)(x-q)=x^2-(p+q)x+n$, como $\varphi(n)= n+1-(p+q)$ tenemos $(x-p)(x-q)=x^2+(\varphi(n)-n-1)x+n$, luego p y q son las raices de este polinomio.

In [38]:
x = var('x')
solve(x^2-(n+1-phi)*x+n, x)

[x == 7789, x == 8009]

Al calcular las raíces de este polinomio se obtienen 7789 y 8009 como p y q.

## Apartado 5

**Intenta factorizar n a partir de e y d.**

Consideramos $k=ed-1=498930432$

In [65]:
k = e*d-1
print(k)

498930432


In [66]:
primos = [2, 3, 5, 7, 11, 13, 17, 19]

for a in primos:
    print(power_mod(a,k,n), a, k ,n)


1 2 498930432 62382101
1 3 498930432 62382101
1 5 498930432 62382101
1 7 498930432 62382101
1 11 498930432 62382101
1 13 498930432 62382101
1 17 498930432 62382101
1 19 498930432 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [67]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 249465216 62382101
1 3 249465216 62382101
1 5 249465216 62382101
1 7 249465216 62382101
1 11 249465216 62382101
1 13 249465216 62382101
1 17 249465216 62382101
1 19 249465216 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [68]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 124732608 62382101
1 3 124732608 62382101
1 5 124732608 62382101
1 7 124732608 62382101
1 11 124732608 62382101
1 13 124732608 62382101
1 17 124732608 62382101
1 19 124732608 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [69]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 62366304 62382101
1 3 62366304 62382101
1 5 62366304 62382101
1 7 62366304 62382101
1 11 62366304 62382101
1 13 62366304 62382101
1 17 62366304 62382101
1 19 62366304 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [70]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 31183152 62382101
1 3 31183152 62382101
1 5 31183152 62382101
1 7 31183152 62382101
1 11 31183152 62382101
1 13 31183152 62382101
1 17 31183152 62382101
1 19 31183152 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [71]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 15591576 62382101
1 3 15591576 62382101
1 5 15591576 62382101
1 7 15591576 62382101
1 11 15591576 62382101
1 13 15591576 62382101
1 17 15591576 62382101
1 19 15591576 62382101


Como $a^k \equiv 1 \mod n$ para $a=2,3,5,7,11,13,17,19$ tomamos $k=k/2$.

In [72]:
k = int(k//2)

for a in primos:
    print(power_mod(a,k,n), a, k ,n)

1 2 7795788 62382101
11909382 3 7795788 62382101
1 5 7795788 62382101
1 7 7795788 62382101
1 11 7795788 62382101
1 13 7795788 62382101
1 17 7795788 62382101
11909382 19 7795788 62382101


Como $3^{7795788} \mod n \equiv 11909382$ calculamos $(n,3^{7795788}-1)=7789$ y $\frac{n}{7789}=8009$ son los factores que buscamos.

In [74]:
n.gcd(power_mod(3, 7795788,n)-1)

7789

In [75]:
n/7789

8009